storage: general GC / intent resolution cleanup

spencerkimball opened the pull request:
The goal with this PR is to move the gc queue to use the intent
resolver's machinery entirely. This is accomplished by widening
the `intentResolver` API so that it can be used by `gcQueue`
without having to synthesize an `EndTransaction` request.

Release notes: none.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/20404)
<!-- Reviewable:end -->

tschottdorf commented:
I'll hold off on this until the predecessor is in. In the meantime though, set a real PR title.

spencerkimball commented:
Yes I'll hold off. Nothing special about this PR. It's really just code movement.

---

Review status: 0 of 5 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-:-L-NIOiHaUZxpfC87gHw:bafc3fz)*
<!-- Sent from Reviewable.io -->

spencerkimball committed:
storage: general GC / intent resolution cleanup

The goal with this PR is to move the gc queue to use the intent
resolver's machinery entirely. This is accomplished by widening
the `intentResolver` API so that it can be used by `gcQueue`
without having to synthesize an `EndTransaction` request.

spencerkimball commented:
@tschottdorf might as well review this while it's still fresh – you've seen it all already in the original PR which has since been closed. This has been rebased on the two changes which have already gone in.

tschottdorf commented:
Did a first pass, but too much going on right now to scrutinize. Will do another take tomorrow, but left some comments.

---

Review status: 0 of 5 files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

*[pkg/storage/gc_queue.go, line 113 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-YeLVwYaG-vjQRxqV7:-L-YeLVwYaG-vjQRxqV8:b-ecun9c) ([raw file](https://github.com/cockroachdb/cockroach/blob/1630f6a57a40167db3bbc8cf63ef7fff280e409b/pkg/storage/gc_queue.go#L113)):*
> // A gcFunc sends the chunks of GCRequests in the order in which they appear. It may but does not have to return on the first error.

---

*[pkg/storage/gc_queue.go, line 114 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-YebOYr1nzaD4MSgAu:-L-YebOYr1nzaD4MSgAv:bdxkhf) ([raw file](https://github.com/cockroachdb/cockroach/blob/1630f6a57a40167db3bbc8cf63ef7fff280e409b/pkg/storage/gc_queue.go#L114)):*
> // A cleanupIntentsFunc synchronously resolves the supplied intents (which may be `PENDING`, in which case they are first pushed) while taking care of proper batching.

---

*[pkg/storage/gc_queue.go, line 115 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-YfFJEd940t4pVKEyl:-L-YfFJFl7MNYLOLBOh7:b-s9l68f) ([raw file](https://github.com/cockroachdb/cockroach/blob/1630f6a57a40167db3bbc8cf63ef7fff280e409b/pkg/storage/gc_queue.go#L115)):*
> ```Go
> type gcFunc func(context.Context, [][]roachpb.GCRequest_GCKey, *GCInfo) error
> type cleanupIntentsFunc func(context.Context, []roachpb.Intent) error
> type cleanupTxnIntentsFunc func(context.Context, *roachpb.Transaction, []roachpb.Intent) error
> ```

// A cleanupTxnIntentsFunc cleans up intents from a transaction record, pushing the transaction first if it is `PENDING`.

This should have `async` in the name.

---

*[pkg/storage/gc_queue_test.go, line 610 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-Yk7YFnA9bUfmIYlRr:-L-Yk7YFnA9bUfmIYlRs:b-83sjo4) ([raw file](https://github.com/cockroachdb/cockroach/blob/1630f6a57a40167db3bbc8cf63ef7fff280e409b/pkg/storage/gc_queue_test.go#L610)):*
> ```Go
> 			orig:       gcTxnAndAC - 1,
> 			newStatus:  -1,
> 			expResolve: true,
> ```

Nice.

---

*[pkg/storage/intent_resolver.go, line 220 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-Ylfd16M2lSZmriF6W:-L-Ylfd16M2lSZmriF6X:b-lfc9u8) ([raw file](https://github.com/cockroachdb/cockroach/blob/1630f6a57a40167db3bbc8cf63ef7fff280e409b/pkg/storage/intent_resolver.go#L220)):*
> ```Go
> 	ir.mu.Unlock()
> 	if len(pushIntents) == 0 {
> 		return []roachpb.Intent(nil), nil
> ```

NYC, but `return nil, nil`.

---

*[pkg/storage/intent_resolver.go, line 272 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-YlxsaDdq5a-w5Sdvz:-L-YlxsaDdq5a-w5Sdw-:b7ex6u4) ([raw file](https://github.com/cockroachdb/cockroach/blob/1630f6a57a40167db3bbc8cf63ef7fff280e409b/pkg/storage/intent_resolver.go#L272)):*
> ```Go
> 		pushee, ok := pushedTxns[intent.Txn.ID]
> 		if !ok {
> 			panic(fmt.Sprintf("no PushTxn response for intent %+v", intent))
> ```

NYC, but `log.Fatalf(ctx, "no PushTxn response for intent %+v\nreqs: %+v", intent, pushReqs))`.

---

*[pkg/storage/intent_resolver.go, line 341 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-Yhr3j49piRsOgs7Nh:-L-Yhr3j49piRsOgs7Ni:b-4d240r) ([raw file](https://github.com/cockroachdb/cockroach/blob/1630f6a57a40167db3bbc8cf63ef7fff280e409b/pkg/storage/intent_resolver.go#L341)):*
> // On success, returns the number of resolved intents. On error, a subset of the intents may have been resolved, but zero will be returned.

---

*[pkg/storage/intent_resolver.go, line 374 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-YlAW7Qlk9IZaUPdMc:-L-YlAW7Qlk9IZaUPdMd:b72fdtl) ([raw file](https://github.com/cockroachdb/cockroach/blob/1630f6a57a40167db3bbc8cf63ef7fff280e409b/pkg/storage/intent_resolver.go#L374)):*
> ```Go
> 		return 0, errors.Wrapf(err, "failed to resolve intents")
> 	}
> 	if pushErr != nil {
> ```

There's *probably* a reason we're only checking `pushErr` here, but it looks scary. Could you make sure that's intended and if not, change it? I wasn't able to find a `return` that made use of this in 

---

*[pkg/storage/intent_resolver.go, line 381 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-Yk_oEQGUioS7syFHR:-L-Yk_oFBJ2RyL6xdAGY:bx5oo0g) ([raw file](https://github.com/cockroachdb/cockroach/blob/1630f6a57a40167db3bbc8cf63ef7fff280e409b/pkg/storage/intent_resolver.go#L381)):*
> // , asynchronously (but returning an error if the intentResolver's semaphore is maxed out).

---

*[pkg/storage/intent_resolver.go, line 390 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-Yi-Bb8EWqE7ppzakk:-L-Yi-Bb8EWqE7ppzakl:b-948j12) ([raw file](https://github.com/cockroachdb/cockroach/blob/1630f6a57a40167db3bbc8cf63ef7fff280e409b/pkg/storage/intent_resolver.go#L390)):*
> ```Go
> 		// dissociate this work from our caller's context and
> 		// timeout.
> 		context.Background(),
> ```

`ir.store.AnnotateCtx(context.Background())` (to get nice logging tags)

---

*[pkg/storage/intent_resolver.go, line 441 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-YkpJ9N-0wDoTiKEYP:-L-YkpJ9N-0wDoTiKEYQ:bcnxh7l) ([raw file](https://github.com/cockroachdb/cockroach/blob/1630f6a57a40167db3bbc8cf63ef7fff280e409b/pkg/storage/intent_resolver.go#L441)):*
> ```Go
> 		}
> 		b := &client.Batch{}
> 		b.AddRawRequest(&roachpb.PushTxnRequest{
> ```

Can't you re-use the existing machinery here?

---

*[pkg/storage/intent_resolver.go, line 454 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-YeCQ-ETGSMqBKwaFQ:-L-YeCQ-ETGSMqBKwaFR:b-jrdxpq) ([raw file](https://github.com/cockroachdb/cockroach/blob/1630f6a57a40167db3bbc8cf63ef7fff280e409b/pkg/storage/intent_resolver.go#L454)):*
> ```Go
> 		}
> 		// Get the pushed txn and update the intents slice.
> 		txn = &b.RawResponse().Responses[0].GetInner().(*roachpb.PushTxnResponse).PusheeTxn
> ```

I don't think you need the `&`.

---

*[pkg/storage/intent_resolver.go, line 507 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-Ykw8OBsxey7rgUJun:-L-Ykw8OBsxey7rgUJuo:b4ji8w7) ([raw file](https://github.com/cockroachdb/cockroach/blob/1630f6a57a40167db3bbc8cf63ef7fff280e409b/pkg/storage/intent_resolver.go#L507)):*
> ```Go
> 	}
> 
> 	return nil
> ```

I'm not sure we actually want to use that pattern, but you could make use of the fact that `errors.Wrapf(nil, anything) == nil`, i.e. just `return errors.Wrapf(ir.store.db.Run(...), "could not GC...")`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-:-L-YmMOMvBOQID9N4Fl4:b-cj95vg)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: 0 of 5 files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

*[pkg/storage/gc_queue.go, line 113 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-YeLVwYaG-vjQRxqV7:-L-Yo0zKLjvOLac1YtHp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1630f6a57a40167db3bbc8cf63ef7fff280e409b/pkg/storage/gc_queue.go#L113)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

> // A gcFunc sends the chunks of GCRequests in the order in which they appear. It may but does not have to return on the first error.
</blockquote></details>

Done.

---

*[pkg/storage/gc_queue.go, line 114 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-YebOYr1nzaD4MSgAu:-L-YoHnU2iidy5iayZ6q:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1630f6a57a40167db3bbc8cf63ef7fff280e409b/pkg/storage/gc_queue.go#L114)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

> // A cleanupIntentsFunc synchronously resolves the supplied intents (which may be `PENDING`, in which case they are first pushed) while taking care of proper batching.
</blockquote></details>

Done.

---

*[pkg/storage/gc_queue.go, line 115 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-YfFJEd940t4pVKEyl:-L-YoHYspLZ0QqpYjFms:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1630f6a57a40167db3bbc8cf63ef7fff280e409b/pkg/storage/gc_queue.go#L115)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

// A cleanupTxnIntentsFunc cleans up intents from a transaction record, pushing the transaction first if it is `PENDING`.

This should have `async` in the name.
</blockquote></details>

Done.

---

*[pkg/storage/intent_resolver.go, line 220 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-Ylfd16M2lSZmriF6W:-L-YoPMDhWvff4MdQHrs:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1630f6a57a40167db3bbc8cf63ef7fff280e409b/pkg/storage/intent_resolver.go#L220)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

NYC, but `return nil, nil`.
</blockquote></details>

Done.

---

*[pkg/storage/intent_resolver.go, line 272 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-YlxsaDdq5a-w5Sdvz:-L-YomYqAYnnJaLoZYq7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1630f6a57a40167db3bbc8cf63ef7fff280e409b/pkg/storage/intent_resolver.go#L272)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

NYC, but `log.Fatalf(ctx, "no PushTxn response for intent %+v\nreqs: %+v", intent, pushReqs))`.
</blockquote></details>

Done.

---

*[pkg/storage/intent_resolver.go, line 341 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-Yhr3j49piRsOgs7Nh:-L-Yov89K-OHdTgRzKRO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1630f6a57a40167db3bbc8cf63ef7fff280e409b/pkg/storage/intent_resolver.go#L341)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

> // On success, returns the number of resolved intents. On error, a subset of the intents may have been resolved, but zero will be returned.
</blockquote></details>

Done.

---

*[pkg/storage/intent_resolver.go, line 374 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-YlAW7Qlk9IZaUPdMc:-L-YrHKdwyORsefaghha:bmfttht) ([raw file](https://github.com/cockroachdb/cockroach/blob/1630f6a57a40167db3bbc8cf63ef7fff280e409b/pkg/storage/intent_resolver.go#L374)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

There's *probably* a reason we're only checking `pushErr` here, but it looks scary. Could you make sure that's intended and if not, change it? I wasn't able to find a `return` that made use of this in 
</blockquote></details>

Yeah, that looks like an anachronism. I can't see how it could ever be non-nil where `resolveIntents` isn't also empty. I suppose at some point in the past we were expecting a potential push error but a non-empty set of intents to still attempt resolution on. I've cleaned it up by moving the `pushErr` check right after the call to `maybePushTransactions`.

---

*[pkg/storage/intent_resolver.go, line 381 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-Yk_oEQGUioS7syFHR:-L-YpAC8HhJPk5oizkl6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1630f6a57a40167db3bbc8cf63ef7fff280e409b/pkg/storage/intent_resolver.go#L381)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

> // , asynchronously (but returning an error if the intentResolver's semaphore is maxed out).
</blockquote></details>

Done.

---

*[pkg/storage/intent_resolver.go, line 390 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-Yi-Bb8EWqE7ppzakk:-L-YpGc7thJcAC4OjOGz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1630f6a57a40167db3bbc8cf63ef7fff280e409b/pkg/storage/intent_resolver.go#L390)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

`ir.store.AnnotateCtx(context.Background())` (to get nice logging tags)
</blockquote></details>

Done.

---

*[pkg/storage/intent_resolver.go, line 441 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-YkpJ9N-0wDoTiKEYP:-L-YpV778rmuy1nbQK72:b-w8458r) ([raw file](https://github.com/cockroachdb/cockroach/blob/1630f6a57a40167db3bbc8cf63ef7fff280e409b/pkg/storage/intent_resolver.go#L441)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Can't you re-use the existing machinery here?
</blockquote></details>

I can't use `maybePushTransactions` without lots of changes. The code duplication here for the request seems minimal. Did you have something else in mind?

---

*[pkg/storage/intent_resolver.go, line 454 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-YeCQ-ETGSMqBKwaFQ:-L-YpS41a9HwmUu9kdpm:bbk50n1) ([raw file](https://github.com/cockroachdb/cockroach/blob/1630f6a57a40167db3bbc8cf63ef7fff280e409b/pkg/storage/intent_resolver.go#L454)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

I don't think you need the `&`.
</blockquote></details>

It is needed because I'm updating the `*roachpb.Transaction` supplied as a parameter to the method. 

---

*[pkg/storage/intent_resolver.go, line 507 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-Ykw8OBsxey7rgUJun:-L-Ypt0ZWhJSvMFTMTgF:b-sneomi) ([raw file](https://github.com/cockroachdb/cockroach/blob/1630f6a57a40167db3bbc8cf63ef7fff280e409b/pkg/storage/intent_resolver.go#L507)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

I'm not sure we actually want to use that pattern, but you could make use of the fact that `errors.Wrapf(nil, anything) == nil`, i.e. just `return errors.Wrapf(ir.store.db.Run(...), "could not GC...")`.
</blockquote></details>

Clever. I would be confused by that in the code so I'll probably leave as is unless you think we should move in that direction in general.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
A few remaining comments, but nothing substantial.

---

Reviewed 3 of 5 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/storage/gc_queue.go, line 123 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-hNgbzbsGu2iAfSQS4:-L-hNgbzbsGu2iAfSQS5:ba7g5hu) ([raw file](https://github.com/cockroachdb/cockroach/blob/7428b33faf55400172256bd286fa41b9eba2b140/pkg/storage/gc_queue.go#L123)):*
> // Once all intents are resolved successfully, removes the transaction record.

---

*[pkg/storage/gc_queue.go, line 737 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-hN67yioiyrCMbi5CY:-L-hN67yioiyrCMbi5CZ:b-8us8o6) ([raw file](https://github.com/cockroachdb/cockroach/blob/7428b33faf55400172256bd286fa41b9eba2b140/pkg/storage/gc_queue.go#L737)):*
> ```Go
> 								TxnMeta: *meta.Txn,
> 							}
> 							infoMu.IntentTxns++
> ```

```go
if meta.Txn.Status == roachpb.PENDING { infoMu.PushTxn += 1 }
```

---

*[pkg/storage/gc_queue.go, line 848 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-hMwd5ulP21UZcdrqj:-L-hMwd5ulP21UZcdrqk:b-hc6i8b) ([raw file](https://github.com/cockroachdb/cockroach/blob/7428b33faf55400172256bd286fa41b9eba2b140/pkg/storage/gc_queue.go#L848)):*
> ```Go
> 	}
> 	infoMu.Lock()
> 	infoMu.PushTxn = len(txnMap)
> ```

This should only be incremented for `PENDING` txns. I think you can just move this up into the loop that scans the kv pairs (left a comment there)

---

*[pkg/storage/intent_resolver.go, line 341 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-Yhr3j49piRsOgs7Nh:-L-hTU3H4-sxetkZa8oO:b-4tuybi) ([raw file](https://github.com/cockroachdb/cockroach/blob/1630f6a57a40167db3bbc8cf63ef7fff280e409b/pkg/storage/intent_resolver.go#L341)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Remove `Returns the number of resolved intents` (is duplicated).

---

*[pkg/storage/intent_resolver.go, line 441 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-YkpJ9N-0wDoTiKEYP:-L-hUBbh4z_qJiPdDEFq:b-o8kio7) ([raw file](https://github.com/cockroachdb/cockroach/blob/1630f6a57a40167db3bbc8cf63ef7fff280e409b/pkg/storage/intent_resolver.go#L441)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

I can't use `maybePushTransactions` without lots of changes. The code duplication here for the request seems minimal. Did you have something else in mind?
</blockquote></details>

No, nevermind. This looks good.

---

*[pkg/storage/intent_resolver.go, line 328 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-hTbyHip_TNPpbKp9b:-L-hTbyHip_TNPpbKp9c:bcdpzbd) ([raw file](https://github.com/cockroachdb/cockroach/blob/7428b33faf55400172256bd286fa41b9eba2b140/pkg/storage/intent_resolver.go#L328)):*
> ```Go
> 		}
> 	} else { // EndTransaction
> 		partialTxn := &roachpb.Transaction{
> ```

What's going on here? Why not use the real thing from `item.Arg` (updated with the first intent)?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-:-L-hW7Zz1GlI5KxOvO3j:bmck6a8)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 5 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-:-L-hp_cR8zI3g1nzJeqI:bnfp4nl)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/storage/gc_queue.go, line 123 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-hNgbzbsGu2iAfSQS4:-L-hzOf87NaDLXqxR2q5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7428b33faf55400172256bd286fa41b9eba2b140/pkg/storage/gc_queue.go#L123)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

> // Once all intents are resolved successfully, removes the transaction record.
</blockquote></details>

Done.

---

*[pkg/storage/gc_queue.go, line 737 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-hN67yioiyrCMbi5CY:-L-hz_LAGoxUkQRHbLhh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7428b33faf55400172256bd286fa41b9eba2b140/pkg/storage/gc_queue.go#L737)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

```go
if meta.Txn.Status == roachpb.PENDING { infoMu.PushTxn += 1 }
```
</blockquote></details>

Done.

---

*[pkg/storage/gc_queue.go, line 848 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-hMwd5ulP21UZcdrqj:-L-hzn_ezQ8QbGcEoHqq:b-qu6n94) ([raw file](https://github.com/cockroachdb/cockroach/blob/7428b33faf55400172256bd286fa41b9eba2b140/pkg/storage/gc_queue.go#L848)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

This should only be incremented for `PENDING` txns. I think you can just move this up into the loop that scans the kv pairs (left a comment there)
</blockquote></details>

Good point. Done.

---

*[pkg/storage/intent_resolver.go, line 341 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-Yhr3j49piRsOgs7Nh:-L-i05jM7WqBokrQjdNd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1630f6a57a40167db3bbc8cf63ef7fff280e409b/pkg/storage/intent_resolver.go#L341)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Remove `Returns the number of resolved intents` (is duplicated).
</blockquote></details>

Done.

---

*[pkg/storage/intent_resolver.go, line 328 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-hTbyHip_TNPpbKp9b:-L-i-E71x7wfW7WiDWrJ:b-rs6emf) ([raw file](https://github.com/cockroachdb/cockroach/blob/7428b33faf55400172256bd286fa41b9eba2b140/pkg/storage/intent_resolver.go#L328)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

What's going on here? Why not use the real thing from `item.Arg` (updated with the first intent)?
</blockquote></details>

Because the value for the txn in `item.Arg` is from before the `EndTransaction` is executed, meaning the status is `PENDING`. This isn't new code, by the way.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
I pulled on that partially-populated transaction a little more and I think I have the right solution now (and that'll be the final thing in this review). If you feel overwhelmed, I wouldn't get in the way of merging this as-is (though an issue would need to be filed), but it's the last wart on what has otherwise become something I'm quite happy about otherwise.

---

Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[pkg/storage/gc_queue.go, line 737 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-hN67yioiyrCMbi5CY:-L-iJjeZ-gt4JlfL0U4_:bo9uenr) ([raw file](https://github.com/cockroachdb/cockroach/blob/7428b33faf55400172256bd286fa41b9eba2b140/pkg/storage/gc_queue.go#L737)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

Oh shoot, now I realize the problem. We don't know the state of the transactions which these intents belong to. In fact that's the point of the push. To try to make sure they're either `ABORTED` or `COMMITTED`. And we really do "push" all of the txns, not just the ones which are still `PENDING`; we just might succeed at some and have noops on others.
</blockquote></details>

Ah, you're right.

---

*[pkg/storage/gc_queue.go, line 848 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-hMwd5ulP21UZcdrqj:-L-iJoknyAjzvbsj015W:b-ssqge0) ([raw file](https://github.com/cockroachdb/cockroach/blob/7428b33faf55400172256bd286fa41b9eba2b140/pkg/storage/gc_queue.go#L848)):*
Not suggesting you change it back, but isn't the code I suggested equivalent because all of these txns are pending? Add a comment

> // All transactions in `txnMap` are `PENDING` and we will push them below.

Also, now `PushTxn` is equivalent to `IntentTxns`. It used to also count pushes to discovered txn records in the local span, but these now feed directly into the metrics. Update the comment on `PushTxn` so that it reads (something like)

// PushTxn is the total number of pushes attempted in this cycle. This will usually be equal to `IntentTxns` since
// pushes to transactions whose record lies in this range (but which are not associated to a remaining intent on it)
// happen asynchronously and cannot be accounted for. However, at the level of the reported metric of the same
// name, these pushes *are* tracked, and thus higher up in the stack generally `PushTxn > IntentTxns`.

---

*[pkg/storage/gc_queue.go, line 122 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-iPYd43Qe9NCbyAONz:-L-iPYd43Qe9NCbyAOO-:b6bundr) ([raw file](https://github.com/cockroachdb/cockroach/blob/02bcedde9c3484353d5d99c19b7d11f4e664019a/pkg/storage/gc_queue.go#L122)):*
> ```Go
> type cleanupIntentsFunc func(context.Context, []roachpb.Intent) error
> 
> // A cleanupTxnIntentsFunc cleans up intents from a transaction
> ```

`asynchronously cleans up`

---

*[pkg/storage/gc_queue.go, line 388 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-iPfFI9LQdJZzGZQMk:-L-iPfFI9LQdJZzGZQMl:b-5mryv5) ([raw file](https://github.com/cockroachdb/cockroach/blob/02bcedde9c3484353d5d99c19b7d11f4e664019a/pkg/storage/gc_queue.go#L388)):*
> ```Go
> 		// resolve, invoke the cleanup function.
> 		if txn.Status == roachpb.PENDING || len(txn.Intents) > 0 {
> 			infoMu.ResolveTotal += len(txn.Intents)
> ```

It's weird that you update `PushTxn` in one fell swoop but then go and do `ResolveTotal` piecemeal here. I'd move both here.

---

*[pkg/storage/intent_resolver.go, line 328 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-hTbyHip_TNPpbKp9b:-L-iLG6xogiwkT19n38l:b-s3yaci) ([raw file](https://github.com/cockroachdb/cockroach/blob/7428b33faf55400172256bd286fa41b9eba2b140/pkg/storage/intent_resolver.go#L328)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

Because the value for the txn in `item.Arg` is from before the `EndTransaction` is executed, meaning the status is `PENDING`. This isn't new code, by the way.
</blockquote></details>

I was suggesting that you take the arg and update it with what's in the first intent's `TxnMeta`.
I'm worried that `cleanupTxnIntents` will change and make use of unset fields. You're pretty close to doing that already, for example we call `txnwait.IsExpired`, except that the state here isn't ever `PENDING`.

I think the right way to handle this is to break up `cleanupTxnIntents` into `cleanupFinishedTxnIntents` (which doesn't ever push, checks that `Status != PENDING`, only needs `TxnMeta`, and would be called here) and a method `forceAbortingPush` which you call when required before (and which needs a whole transaction). In fact, I *think* you could just inline `forceAbortingPush` into what is right now `cleanupTxnIntentsOnGCAsync`.

---

*[pkg/storage/intent_resolver.go, line 386 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-iQnidRJfM9AidB_Ft:-L-iQnidRJfM9AidB_Fu:bukdbzg) ([raw file](https://github.com/cockroachdb/cockroach/blob/02bcedde9c3484353d5d99c19b7d11f4e664019a/pkg/storage/intent_resolver.go#L386)):*
> ```Go
> // intentResolver's semaphore is maxed out). This method updates the
> // metrics for intents resolved on GC on success.
> func (ir *intentResolver) cleanupTxnIntentsOnGCAsync(
> ```

It's a mouthful, but `cleanupPendingOrFinishedTxnOnGCAsync`? I'm suggesting that because it's never that clear which state transactions need to be in to be eligible.

---

*[pkg/storage/intent_resolver.go, line 404 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-iQiAxNxHPAFbBHZi0:-L-iQiAxNxHPAFbBHZi1:bxkek44) ([raw file](https://github.com/cockroachdb/cockroach/blob/02bcedde9c3484353d5d99c19b7d11f4e664019a/pkg/storage/intent_resolver.go#L404)):*
> ```Go
> 		false, /* wait */
> 		func(ctx context.Context) {
> 			if err := ir.cleanupTxnIntents(ctx, txn, intents, now); err != nil {
> ```

This would call the force-push first if state is `PENDING`.

---

*[pkg/storage/intent_resolver.go, line 437 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-iRWogFOWbb1OrMCB4:-L-iRWogFOWbb1OrMCB5:byrom54) ([raw file](https://github.com/cockroachdb/cockroach/blob/02bcedde9c3484353d5d99c19b7d11f4e664019a/pkg/storage/intent_resolver.go#L437)):*
> ```Go
> 	}()
> 
> 	// If the transaction is still pending, but expired, push it
> ```

When called from the GC queue, I think you want to bump the `GCPushTxn` GCInfo metric if there is one. You'd have a natural place in `cleanupTxnIntentsOnGCAsync` when the push moves there.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-:-L-iRlTnHIJQpneTz1lI:b-dmtikg)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[pkg/storage/gc_queue.go, line 848 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-hMwd5ulP21UZcdrqj:-L05R_JtaIBrKhbFcr_N:ban8j9a) ([raw file](https://github.com/cockroachdb/cockroach/blob/7428b33faf55400172256bd286fa41b9eba2b140/pkg/storage/gc_queue.go#L848)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Not suggesting you change it back, but isn't the code I suggested equivalent because all of these txns are pending? Add a comment

> // All transactions in `txnMap` are `PENDING` and we will push them below.

Also, now `PushTxn` is equivalent to `IntentTxns`. It used to also count pushes to discovered txn records in the local span, but these now feed directly into the metrics. Update the comment on `PushTxn` so that it reads (something like)

// PushTxn is the total number of pushes attempted in this cycle. This will usually be equal to `IntentTxns` since
// pushes to transactions whose record lies in this range (but which are not associated to a remaining intent on it)
// happen asynchronously and cannot be accounted for. However, at the level of the reported metric of the same
// name, these pushes *are* tracked, and thus higher up in the stack generally `PushTxn > IntentTxns`.

</blockquote></details>

OK, make some changes which I think address your points.

---

*[pkg/storage/gc_queue.go, line 122 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-iPYd43Qe9NCbyAONz:-L04xLBrgOQjHpY2XssX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/02bcedde9c3484353d5d99c19b7d11f4e664019a/pkg/storage/gc_queue.go#L122)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

`asynchronously cleans up`
</blockquote></details>

Done.

---

*[pkg/storage/gc_queue.go, line 388 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-iPfFI9LQdJZzGZQMk:-L04y-gaPdL-jPbiHHrJ:bfd9fjg) ([raw file](https://github.com/cockroachdb/cockroach/blob/02bcedde9c3484353d5d99c19b7d11f4e664019a/pkg/storage/gc_queue.go#L388)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

It's weird that you update `PushTxn` in one fell swoop but then go and do `ResolveTotal` piecemeal here. I'd move both here.
</blockquote></details>

I moved these into the async code in `intentResolver`, which is where we're updating `PushTxn` as well as `ResolveSuccess`.

---

*[pkg/storage/intent_resolver.go, line 328 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-hTbyHip_TNPpbKp9b:-L050sxKTFEwZencc2kU:batqtpi) ([raw file](https://github.com/cockroachdb/cockroach/blob/7428b33faf55400172256bd286fa41b9eba2b140/pkg/storage/intent_resolver.go#L328)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

I was suggesting that you take the arg and update it with what's in the first intent's `TxnMeta`.
I'm worried that `cleanupTxnIntents` will change and make use of unset fields. You're pretty close to doing that already, for example we call `txnwait.IsExpired`, except that the state here isn't ever `PENDING`.

I think the right way to handle this is to break up `cleanupTxnIntents` into `cleanupFinishedTxnIntents` (which doesn't ever push, checks that `Status != PENDING`, only needs `TxnMeta`, and would be called here) and a method `forceAbortingPush` which you call when required before (and which needs a whole transaction). In fact, I *think* you could just inline `forceAbortingPush` into what is right now `cleanupTxnIntentsOnGCAsync`.
</blockquote></details>

OK, took your suggestions and reworked this.

---

*[pkg/storage/intent_resolver.go, line 386 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-iQnidRJfM9AidB_Ft:-L05PE81dTugcSCCaswp:b-i659xa) ([raw file](https://github.com/cockroachdb/cockroach/blob/02bcedde9c3484353d5d99c19b7d11f4e664019a/pkg/storage/intent_resolver.go#L386)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

It's a mouthful, but `cleanupPendingOrFinishedTxnOnGCAsync`? I'm suggesting that because it's never that clear which state transactions need to be in to be eligible.
</blockquote></details>

The renaming strikes me as unnecessary. I updated the comment instead. Let me know if that works for you.

---

*[pkg/storage/intent_resolver.go, line 404 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-iQiAxNxHPAFbBHZi0:-L05POtjf4TVoeJamGkO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/02bcedde9c3484353d5d99c19b7d11f4e664019a/pkg/storage/intent_resolver.go#L404)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

This would call the force-push first if state is `PENDING`.
</blockquote></details>

Done.

---

*[pkg/storage/intent_resolver.go, line 437 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-iRWogFOWbb1OrMCB4:-L05PvfjBdk6zlCyPfpS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/02bcedde9c3484353d5d99c19b7d11f4e664019a/pkg/storage/intent_resolver.go#L437)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

When called from the GC queue, I think you want to bump the `GCPushTxn` GCInfo metric if there is one. You'd have a natural place in `cleanupTxnIntentsOnGCAsync` when the push moves there.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> mod some insubstantial comments!

---

Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/storage/intent_resolver.go, line 386 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L-iQnidRJfM9AidB_Ft:-L05ZLxGUedY32iNcB-L:b4c9lyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/02bcedde9c3484353d5d99c19b7d11f4e664019a/pkg/storage/intent_resolver.go#L386)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary><blockquote>

The renaming strikes me as unnecessary. I updated the comment instead. Let me know if that works for you.
</blockquote></details>

:thumbsup: 

---

*[pkg/storage/intent_resolver.go, line 425 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L05Za9F8ZKjWGDBms1a:-L05Za9F8ZKjWGDBms1b:bbeze1m) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e18ab95e3a70c8010f59cd89a6e2e401d11e24a/pkg/storage/intent_resolver.go#L425)):*
> ```Go
> 		false, /* wait */
> 		func(ctx context.Context) {
> 			if locked, release := ir.lockInFlightTxnCleanup(ctx, txn.ID); locked {
> ```

nit: de-denting improves readability in this case, so

```go
locked, release := ir.lockInFlightTxnCleanup(ctx, txn.ID)
if !locked {
  return
}
defer release()

// work

```

---

*[pkg/storage/intent_resolver.go, line 449 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L05_5jqcVBKg8wbxPlm:-L05_5jqcVBKg8wbxPln:b-ob8k4o) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e18ab95e3a70c8010f59cd89a6e2e401d11e24a/pkg/storage/intent_resolver.go#L449)):*
> ```Go
> 						return
> 					}
> 					ir.store.metrics.GCPushTxn.Inc(1)
> ```

We track the number of *attempted* pushes, so this should move before the call to `Run` above.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-:-L05_GUDNqEMQDGcEi_V:bc3y5cn)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/storage/intent_resolver.go, line 425 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L05Za9F8ZKjWGDBms1a:-L06w0yFOFvhQfGFoh2j:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e18ab95e3a70c8010f59cd89a6e2e401d11e24a/pkg/storage/intent_resolver.go#L425)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

nit: de-denting improves readability in this case, so

```go
locked, release := ir.lockInFlightTxnCleanup(ctx, txn.ID)
if !locked {
  return
}
defer release()

// work

```
</blockquote></details>

Done.

---

*[pkg/storage/intent_resolver.go, line 449 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-L05_5jqcVBKg8wbxPlm:-L06w63dvueZiiolM-K5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e18ab95e3a70c8010f59cd89a6e2e401d11e24a/pkg/storage/intent_resolver.go#L449)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

We track the number of *attempted* pushes, so this should move before the call to `Run` above.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> merge away and thanks!

---

Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20404#-:-L06wx9aIkF5Yk0KD1Uq:b-3guf3g)*
<!-- Sent from Reviewable.io -->
